Rearranged installation process

alexiswl opened the pull request:
## Dockerfile changes
  * Moved from ubuntu xenial to focal (16.04 -> 20.04)
  * Moved COPY commands to top of Dockerfile
  * Added ARG list:
    * TARGETPLATFORM, BUILDPLATFORM for multiplatform builds
    * USER, UID, GID, GROUP for non root user access
    * MAMABAFORGE_VERSION and PYPY version for downloading specific miniforge
    * CRAN_KEY_SERVER and CRAN_RECV_KEY for R installation through apt

  * Moved RUN into one for more efficient container 
  * Removed R packages from conda and install through Rscript instead
  
## TODO list

- [ ] ~~Try multiplatform build over gravitron / intel instances on AWS~~ On hold, waiting for arm vms from GitHub actions
- [x] Test out non-privileged components of dockerfile
- [x] Test compatibility with 'apt' installation of R with conda commands
- [x] ~~Conda version of awscli is still version 1~~ See [this error](https://github.com/aws/aws-cli/issues/4947)
- [x] Hush conda installations
- [x] Some R packages fail to install - reduce verbosity on R installations in order to find them
- [x] Fix R installation error `No package 'libcurl' found` causes failure in R installation
- [x] Fix R installation error `Configuration failed because openssl was not found -> Install `libssl-dev`
- [x] Fix BioR installation errors `Configuration failed because libcurl was not found` -> Install `libcurl4-openssl-dev`
- [x] `Warning message: packages ‘dt’, ‘matrixstats’, ‘rcircos’ are not available for this version of R`
- [x] Pandoc should probably be installed through source / `deb` see [here](https://github.com/jgm/pandoc/releases/tag/2.14.1)
- [x] Add `CMD` to end of the Dockerfile

alexiswl commented:
@JMarzec and @skanwal 

As a potential solution to the HPC / Cloud discrepancy, I've got this Dockerfile approved for use on HPC at `/g/data/gx8/projects/Lucattini/singularity_boxes/rnasum_4.0--dev.sif` if you wanted to have a play around:

### Getting started:

1. Enter a interactive qsub node:
```bash
$ qsub -I -P gx8 -q copyq -l walltime=10:00:00,ncpus=1,mem=10G,storage=gdata/gx8+scratch/gx8
```

2. Load the singularity module (v3.7.0)
```bash
$ module load singularity
$ singularity --version
singularity version 3.7.0
```

### Looking around

1. Inspect the container:
```bash
$ singularity inspect /g/data/gx8/projects/Lucattini/singularity_boxes/rnasum_4.0--dev.sif
org.label-schema.build-arch: amd64
org.label-schema.build-date: Tuesday_31_August_2021_2:26:20_UTC
org.label-schema.schema-version: 1.0
org.label-schema.usage.singularity.deffile.bootstrap: docker
org.label-schema.usage.singularity.deffile.from: umccr/rnasum:4.0--dev
org.label-schema.usage.singularity.version: 3.8.2
```

2. Enter the container and have a look around:

2a. Log in to the shell
```bash
$ singularity shell /g/data/gx8/projects/Lucattini/singularity_boxes/rnasum_4.0--dev.sif
```

2b.  Have a look at the environment `PATH`
```
Singularity> echo $PATH
/miniconda/envs/rnasum/bin:/miniconda/bin:/miniconda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

2c. Are our report files where we expect them?

```
Singularity> ls /rmd_files/
RNAseq_report.R  RNAseq_report.Rmd
```


2d. Do we have access to our data? Yes!
```
Singularity> ls /g/data/gx8/projects/ -d
/g/data/gx8/projects/
```

2e What about our .R files?
```
Singularity> ls -la | grep '\.R'
-rw-rw-r--   1 al9223 gx8         318 Sep  2 14:36 .RData
-rw-rw-r--   1 al9223 gx8          54 Sep  2 14:32 .Renviron
-rw-------   1 al9223 gx8          38 Sep  2 14:36 .Rhistory
-rw-rw-r--   1 al9223 gx8        1104 Jan 16  2020 .Rprofile
```

3 Using R (we're still inside the singularity shell)
```
Singularity> R
R version 4.1.1 (2021-08-10) -- "Kick Things"
....
Type 'q()' to quit R.

3a. Check out our libpaths:

Note this first path is due to my `.Renviron` and `.Rprofile` in my home directory:
> .libPaths()
[1] "/g/data/gx8/extras/Lucattini/R/4.1.1/library"
[2] "/usr/local/lib/R/site-library"
[3] "/usr/lib/R/site-library"
[4] "/usr/lib/R/library"
```

3b. Can we install a package? You betcha
```
> install.packages("renv")
Installing package into ‘/g/data/gx8/extras/Lucattini/R/4.1.1/library’
(as ‘lib’ is unspecified)
trying URL 'https://cran.ms.unimelb.edu.au/src/contrib/renv_0.14.0.tar.gz'
Content type 'application/octet-stream' length 448230 bytes (437 KB)
==================================================
downloaded 437 KB

* installing *source* package ‘renv’ ...
** package ‘renv’ successfully unpacked and MD5 sums checked
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (renv)

The downloaded source packages are in
        ‘/jobfs/27678272.gadi-pbs/Rtmp694E84/downloaded_packages’
> quit()

# And here's the package!
Singularity> ls "/g/data/gx8/extras/Lucattini/R/4.1.1/library/renv/" -d
/g/data/gx8/extras/Lucattini/R/4.1.1/library/renv/
```

3c: Now exit the singularity shell:
```
Singularity> exit
```

4. Running a script through singularity

One can use the `singularity exec` command for this:

```bash
$ singularity exec /g/data/gx8/projects/Lucattini/singularity_boxes/rnasum_4.0--dev.sif Rscript /rmd_files/RNAseq_report.R

Please type in required arguments!


command example:

Rscript RNAseq_report.R  --sample_name test_sample_WTS  --dataset PAAD  --bcbio_rnaseq $(pwd)/../data/test_data/final/test_sample_WTS  --report_dir $(pwd)/../data/test_data/final/test_sample_WTS/RNAseq_report

or

Rscript RNAseq_report.R  --sample_name test_sample_WTS  --dataset PAAD  --dragen_rnaseq $(pwd)/../data/test_data/stratus/test_sample_WTS  --report_dir $(pwd)/../data/test_data/stratus/test_sample_WTS/RNAseq_report
```

Hmmm, okay other than needing arguments, we probably need to use the `--vanilla` Rscript so our `.Rprofile` and `.Renviron` don't affect the container. We can also use the `--contain` singularity parameter in the singularity exec function to ensure that our home directory is not mounted.  

Adding custom R libraries. We can use the `--vanilla` flag in conjunction with the `R_LIBS_USER` environment variable to ensure we have both the right internal and external packages without changing our `.Rprofile` or `.Renviron` files.


```
$ mkdir -p "$HOME/R/singularity/rnasum/4.0--dev/R/library"
$ singularity exec --env R_LIBS_USER="$HOME/R/singularity/rnasum/4.0--dev/R/library" /g/data/gx8/projects/Lucattini/singularity_boxes/rnasum_4.0--dev.sif Rscript /rmd_files/RNAseq_report.R

Please type in required arguments!


command example:

Rscript RNAseq_report.R  --sample_name test_sample_WTS  --dataset PAAD  --bcbio_rnaseq $(pwd)/../data/test_data/final/test_sample_WTS  --report_dir $(pwd)/../data/test_data/final/test_sample_WTS/RNAseq_report

or

Rscript RNAseq_report.R  --sample_name test_sample_WTS  --dataset PAAD  --dragen_rnaseq $(pwd)/../data/test_data/stratus/test_sample_WTS  --report_dir $(pwd)/../data/test_data/stratus/test_sample_WTS/RNAseq_report
```

Let me know if you have any questions!

alexiswl commented:
Looking at renv as a way to version control the builds. Will send through a process running renv::update() as a method to update the packages if one wishes to have the latest packages for a new release.

alexiswl commented:
`renv` summary: 

# Update all packages: 

One can use the `renv::update()` command to update all packages in the docker image.  

This will update the lock file meaning when the container is rebuilt, new versions of the packages will be installed. 

```bash
cd rnasum;
docker run --rm \
  ghcr.io/umccr/rnasum:v0.4.2-dev \
  sh -c 'Rscript -e "renv::update(); \
                             renv::snapshot()" >/dev/null 2>&1 && \
           cat /rmd_files/renv.lock' > rmd_files/renv.lock
```

## Show difference in package updates (good for change logs) via renv.lock
```bash
cd rnasum;

dyff between --omit-header \
  <(docker run --rm \
        ghcr.io/umccr/rnasum:v0.4.2-dev \
          cat renv.lock) \
  <(docker run --rm \
        ghcr.io/umccr/rnasum:v0.4.2-dev \
        sh -c 'Rscript -e "renv::update(); \
                                   renv::snapshot()" >/dev/null 2>&1 && \
                  cat renv.lock')
```


# Update a single package

Just set the packages command to either a string or list

```bash
cd rnasum;
docker run --rm \
  ghcr.io/umccr/rnasum:v0.4.2-dev \
  sh -c 'Rscript -e "renv::update(packages=\"DT\"); \
                             renv::snapshot()" >/dev/null 2>&1 && \
           cat /rmd_files/renv.lock' > rmd_files/renv.lock
```

## Show renv.lock changes

```bash
cd rnasum;

dyff between --omit-header \
  <(docker run --rm \
        ghcr.io/umccr/rnasum:v0.4.2-dev \
          cat renv.lock) \
  <(docker run --rm \
        ghcr.io/umccr/rnasum:v0.4.2-dev \
        sh -c 'Rscript -e "renv::update(packages=\"DT\"); \
                                   renv::snapshot()" >/dev/null 2>&1 && \
                  cat renv.lock')
```

Which will yield something like:

```
Packages
  + one map entry added:
    BiocVersion:
    │ Hash: e57437f82a5c13263a9cf442b9796ba3
    │ Package: BiocVersion
    │ Requirements: []
    │ Source: Bioconductor
    │ Version: 3.14.0
    │ git_branch: master
    │ git_last_commit: aa56d93
    │ git_last_commit_date: 2021-05-19
    │ git_url: "https://git.bioconductor.org/packages/BiocVersion"

Packages.DT.Hash
  ± value change
    - 45fa28dbf288cd606e13ca35d3d72437
    + fc53164de9ee32692eef6fdd14115381

Packages.DT.Version
  ± value change
    - 0.21
    + 0.22
```

pdiakumis commented:
@alexiswl I think only `manhattanly` needs to be added to the renv lock file and the Docker image needs to be rebuilt; the rest of the conda deps are listed okay (see `trn.be7f009b2bbd4de28976f9e5eb0842bd` stderr).

alexiswl commented:
Thanks @pdiakumis well discovered, wondering why we didn't see this before? 

Added in https://github.com/umccr/RNAsum/pull/58.  

We shouldn't have to update the cwl repo since the docker file will be just updated with the same tag

pdiakumis commented:
It's in the conda env, just not the renv lock, which I think is the new way the Docker container is set up - https://github.com/umccr/RNAsum/pull/45/files#diff-66977b0a681cfc0549861da12dc76bda9b357c2bb33ca245233514f123baea60L50

skanwal commented:
> On hold, waiting for arm vms from GitHub actions

 @alexiswl - Currently, building the image locally for testing purposes and while reading through comments wondering if the above point means it's not yet possible to test the updated image locally?

alexiswl commented:
> @alexiswl - Currently, building the image locally for testing purposes and while reading through comments wondering if the above point means it's not yet possible to test the updated image locally?

Yes that's correct, you might need to spin up an ec2 instance in the meantime.  

Don't think there's any docker images out there for arm that support R4 without having to install it from source.

alexiswl commented:
But this is where an install.sh script could come in handy, so long as you have the pre-reqs installed, you could then install RNASum locally without docker

alexiswl commented:
> It's in the conda env, just not the renv lock, which I think is the new way the Docker container is set up - https://github.com/umccr/RNAsum/pull/45/files#diff-66977b0a681cfc0549861da12dc76bda9b357c2bb33ca245233514f123baea60L50

I've added it into the `renv_required_libraries.R` file so it's picked up the next time.  
Needed to use the 'hydrate' command from renv so its added to the lock if the lock already exists.  

This will be useful for when we just want to add in a library into the RMD file set but don't want to have to run any of the renv commands.  

I've also added in an extra job into the GH actions workflow that reupdates the renv.lock file if it was changed by the renv::hydrate command. 

Recompiling, so might need to wait for a rebuild to check if its good to go know

skanwal commented on envm/Dockerfile:
```diff
@@ -1,46 +1,325 @@
+# See https://github.com/rstudio/renv/issues/934 for more details
+# We've also added in rmd_files/renv_required_libraries.R
+# To make sure that libraries that aren't auto downloaded by renv are added to the list
+RUN echo "Install Bioconductor packages thorugh R" 1>&2 && \
```
typo in echo command "through"

skanwal commented on envm/Dockerfile:
```diff
@@ -1,46 +1,325 @@
-    apt-get install -y tzdata && \
-    echo "Etc/UTC" > /etc/timezone && \
-    dpkg-reconfigure -f noninteractive tzdata
+# Biocmanager pacakges that need to be installed prior to renv being activated
```
typo in comment "packages"

skanwal commented on envm/Dockerfile:
```diff
@@ -1,46 +1,325 @@
+RUN echo "Install Bioconductor packages thorugh R" 1>&2 && \
+      Rscript -e 'options(repos=BiocManager::repositories()); \
+                  BiocManager::install(ask=FALSE, pkgs=c( \
+                                                         "EDASeq", \
+                                                         "limma", \
+                                                         "preprocessCore", \
+                                                         "edgeR", \
+                                                         "rhdf5", \
+                                                         "tximport", \
+                                                         "EnsDb.Hsapiens.v75", \
+                                                         "EnsDb.Hsapiens.v86", \
+                                                         "BSgenome.Hsapiens.UCSC.hg19", \
+                                                         "BSgenome.Hsapiens.UCSC.hg38" ), \
```
are these the only renv dependencies? Couldn't find this info in the issue linked.

alexiswl replied:
The dependencies that aren't explicitly specified are in [this file](https://github.com/umccr/RNAsum/blob/dockerfile-updates/rmd_files/renv_required_libraries.R). The issue linked likely doesn't correlate to the current issue. The reasons these packages needed to be installed locally was due to the way renv was trying to install them. I can try and emulate this at some point so its resolved and we don't have to specify many of the Bioc packages first.

skanwal replied:
Okay, to summarise - there are couple dependencies that are installed via mamba (plotly-orca, libopenblas==0.3.7) and renv dependencies are installed separately as listed in this [file](https://github.com/umccr/RNAsum/blob/dockerfile-updates/rmd_files/renv_required_libraries.R). 
Why are these dependencies (line 233-242) not part of the linked file?

skanwal replied:
> The installation of bioconductor packages inside the dockerfile are for packages that renv has trouble installing.

I'd add this as a comment to docker file. So that next time we hit a package that renv is having trouble installing we can add to this bioconductor installation command. 

> They need to be inside the rmd_files directory
What is `they` here?

skanwal commented on envm/Dockerfile:
```diff
@@ -1,46 +1,325 @@
+        --name base \
+        --channel conda-forge \
+        mamba && \
+    echo "Install environment with mamba" 1>&2 && \
+      /miniforge/bin/mamba env create --quiet \
+        --file /tmp/environment.yaml && \
```
Why is mamba being used if we are later using renv to create the environment?

alexiswl replied:
There are a couple of dependencies in the conda env that aren't R related which is why this is still there.

```
dependencies:
        - plotly-orca
        - libopenblas==0.3.7
```

One could try install these through other means if we want to remove conda entirely.

alexiswl commented:
This should be good to go now
